Issue-103 - HEAD must have same headers as GET, except payload header fields

zimeon opened the pull request:
Fixes #103. Discussed on API call today: We need HEAD to include `Digest` header so that persistence fixity works and simply strengthening RFC7231's SHOULD to a MUST with the same payload header exemption seems like the best approach.

zimeon committed:
HEAD must include all non-payload header fields

ajs6f commented:
Why was a blanket approach like this considered better than a narrowly-targeted approach (i.e. requiring the operation of `Want-Digest`)?

acoburn commented:
I will point out that anyone using JAX-RS and Jersey will have a _very_ hard time conforming to this. That includes the current Modeshape-based implementation as well as (probably) any other java-based implementation. Jersey, for instance, will return `Content-Length: 0` on a `HEAD` request, which definitely isn't the same as what is returned with `GET`, and while this language allows that header to be _omitted_ under `HEAD`, it doesn't allow it to be different.

acoburn commented:
For the reasons described above and in #103 I am 👎  on this change. There are good reasons that RFC7231 uses `SHOULD`. Strengthening it to `MUST` is shortsighted. If the goal is to make sure that `Want-Digest` is supported by `HEAD` requests, I would suggest limiting the text in this specification to that.

zimeon commented:
@acoburn `Content-Length` is a payload header and thus exempted per https://tools.ietf.org/html/rfc7231#section-3.3 . IMO returning `Content-Lenght: 0` for HEAD when the real length is non-zero is just broken.

ajs6f commented:
@zimeon https://tools.ietf.org/html/rfc7231#section-4.3.2 makes clear that payload headers may be _omitted_. It does not permit them to be present, but different from what they would be for a `GET`. @acoburn is raising a practical objection.

awoods commented:
I have been doing some experimenting and have determined that in the absence of an application-provided Content-Length header, the servlet container (Jetty, Tomcat, etc) chooses to set the Content-Length or not on HEAD requests.
For example, 
- I use Tomcat8 locally, and it omits the Content-Length header on HEAD
- The UT-Austin demo server (http://demo.fcrepo.org:8080/fcrepo/) uses Tomcat7 and it returns Content-Length=0 on HEAD
- The Jetty server used with Fedora's maven-jetty plugin also returns Content-Length=0 on HEAD

The point here is that there does not appear to be anything we can do at the Fedora level to dictate the servlet container's behavior.
As such, calling out `Digest` is probably the right path forward.

Related Tomcat bug:
https://bz.apache.org/bugzilla/show_bug.cgi?id=59310#c8

zimeon commented:
Thanks @acoburn & @awoods -- looks we should just follow RFC7231 except for a MUST for `Digest`. Will update the PR tomorrow.

zimeon committed:
Update to MUST for Digest only

zimeon commented:
I have updated to say MUST only for `Digest`. Although it differs from the writing style of RFC7231 I also changed "header field(s)" to "header(s)" in the three places it occurs in this section, in order to be consistent with the rest of the spec
